macaroonbakery/httpbakery/client.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2017 Canonical Ltd.
+# Licensed under the LGPLv3, see LICENCE file for details.
+
+import base64
+try:
+    from cookielib import Cookie, CookieJar         # Python 2
+except ImportError:
+    from http.cookiejar import Cookie, CookieJar    # Python 3  # NOQA
+import requests
+try:
+    from urlparse import urlparse                   # Python 2
+    from urlparse import urljoin                   # Python 2
+except ImportError:
+    # from urllib.parse import urlparse               # Python 3  # NOQA
+    from urllib.parse import urlparse               # Python 3  # NOQA
+    from urllib.parse import urljoin               # Python 3  # NOQA
+import webbrowser
+
+from macaroonbakery.bakery import discharge_all
+from macaroonbakery import utils
+
+ERR_INTERACTION_REQUIRED = 'interaction required'
+ERR_DISCHARGE_REQUIRED = 'macaroon discharge required'
+TIME_OUT = 30
+MAX_DISCHARGE_RETRIES = 3
+
+
+class BakeryAuth:
+    ''' BakeryAuth holds the context for making HTTP requests
+        that automatically acquire and discharge macaroons around the requests
+        framework.
+        Usage:
+            from macaroonbakery import httpbakery
+            jar = requests.cookies.RequestsCookieJar()
+            resp = requests.get('some protected url',
+                                auth=httpbakery.BakeryAuth(cookies=jar))
+            resp.raise_for_status()
+    '''
+    def __init__(self, visit_page=None, key=None, cookies=None):
+        '''
+
+        @param visit_page: holds a Visitor that is called when the
+        discharge process requires further interaction.
+        @param key: holds the client's key. If set, the client will try to
+        discharge third party caveats with the special location "local" by
+        using this key.
+        @param cookies: storage for the cookies, that will or contains a cookie
+        for a given URL on the given cookie jar that will holds the macaroon
+        slice.
+        '''
+        if visit_page is None:
+            visit_page = _visit_page_with_browser
+        if cookies is None:
+            cookies = requests.cookies.RequestsCookieJar()
+        if 'agent-login' in cookies.keys():
+            self._visit_page = _visit_page_for_agent(cookies, key)
+        else:
+            self._visit_page = visit_page
+        self._jar = cookies
+        self._key = key
+
+    def __call__(self, req):
+        req.headers['Bakery-Protocol-Version'] = '1'
+        hook = _prepare_discharge_hook(req.copy(), self._key, self._jar,
+                                       self._visit_page)
+        req.register_hook(event='response', hook=hook)
+        return req
+
+
+def _prepare_discharge_hook(req, key, jar, visit_page):
+
+    class Retry:
```

Comment:
Maybe  a comment might be justified:

    # Define a local class so that can use its class variable as
    # mutable state accessed by the closures below.

?